File: crates/SphereWebAudioCore/src/dsp/resample.rs
//! Linear-interpolation resampler.
//!
//! Convention (matches the TypeScript resample.ts):
//!   speed_ratio 2.0 → output has half the samples → plays twice as fast.
//!   speed_ratio 0.5 → output has double the samples → plays at half speed.
//!
//! out_len = ceil(in_len / speed_ratio)

pub fn resample_linear(input: &[f32], speed_ratio: f32) -> Vec<f32> {
    let ratio = speed_ratio.clamp(0.25, 4.0);
    if input.is_empty() {
        return Vec::new();
    }
    if (ratio - 1.0).abs() < 1e-6 {
        return input.to_vec();
    }

    let out_len = ((input.len() as f32) / ratio).ceil() as usize;
    let out_len = out_len.max(1);
    let mut output = vec![0.0_f32; out_len];
    let last_idx = input.len() - 1;

    for i in 0..out_len {
        let src_pos = i as f32 * ratio;
        let lo = (src_pos.floor() as usize).min(last_idx);
        let hi = (lo + 1).min(last_idx);
        let frac = src_pos - lo as f32;
        output[i] = input[lo] + (input[hi] - input[lo]) * frac;
    }

    output
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_input_safe() {
        assert!(resample_linear(&[], 2.0).is_empty());
    }

    #[test]
    fn speed_2_halves_length() {
        let input: Vec<f32> = (0..100).map(|i| i as f32).collect();
        let out = resample_linear(&input, 2.0);
        assert_eq!(out.len(), 50);
    }

    #[test]
    fn speed_half_doubles_length() {
        let input: Vec<f32> = (0..100).map(|i| i as f32).collect();
        let out = resample_linear(&input, 0.5);
        assert_eq!(out.len(), 200);
    }

    #[test]
    fn no_nan_or_inf() {
        let input: Vec<f32> = (0..256).map(|i| (i as f32).sin()).collect();
        for val in resample_linear(&input, 1.5) {
            assert!(val.is_finite(), "output contains non-finite value");
        }
    }
}

File: apps/web/src/audio/dsp/resample.ts
// Use the wide typed-array form so inputs from both AudioBuffer and new Float32Array() are accepted.
type F32 = Float32Array<ArrayBufferLike>;

/**
 * Linear interpolation resampler.
 *
 * Convention:
 *   speedRatio 2.0 → output is half the length → audio plays twice as fast → higher pitch
 *   speedRatio 0.5 → output is double the length → audio plays at half speed → lower pitch
 *
 * outLength = ceil(inLength / speedRatio)
 */
export function resampleLinear(input: F32, speedRatio: number): Float32Array {
  const ratio = Math.max(0.25, Math.min(4.0, speedRatio));

  if (input.length === 0) return new Float32Array(0);
  if (ratio === 1) return new Float32Array(input);

  const outLen = Math.max(1, Math.ceil(input.length / ratio));
  const output = new Float32Array(outLen);
  const lastIdx = input.length - 1;

  for (let i = 0; i < outLen; i++) {
    const srcPos = i * ratio;
    const lo = Math.floor(srcPos) | 0;
    const hi = lo < lastIdx ? lo + 1 : lastIdx;
    const frac = srcPos - lo;
    output[i] = input[lo] + (input[hi] - input[lo]) * frac;
  }

  return output;
}

/** Apply resampleLinear to each channel. */
export function resampleChannels(channels: F32[], speedRatio: number): Float32Array[] {
  return channels.map((ch) => resampleLinear(ch, speedRatio));
}
